tests/saitests/py3/sai_qos_tests.py

```diff
@@ -2881,27 +2883,32 @@ def runTest(self):
                     print(
                         "Verifying XOFF has now been triggered on final iteration", file=sys.stderr)
                     sys.stderr.flush()
-                    time.sleep(8)
+                    time.sleep(4)
                     recv_counters = sai_thrift_read_port_counters(
                         self.src_client, self.asic_type, port_list['src'][src_port_id])[0]
                     xoff_txd = recv_counters[self.pg_cntr_indices[i]] - \
-                        recv_counters_bases[i][self.pg_cntr_indices[i]]
+                        recv_counters_bases[src_port_id][self.pg_cntr_indices[i]]
                     assert xoff_txd > 0, "Failed to trigger XOFF on final iteration"
 
             # Verify no ingress/egress drops for all ports
-            recv_counters_list = [sai_thrift_read_port_counters(self.src_client, self.asic_type, port_list['src'][sid])[
-                0] for sid in self.src_port_ids]
-            xmit_counters_list = [sai_thrift_read_port_counters(self.dst_client, self.asic_type, port_list['dst'][sid])[
-                0] for sid in self.dst_port_ids]
-            for i in range(len(self.src_port_ids)):
-                for cntr in self.ingress_counters:
-                    drops = recv_counters_list[i][cntr] - \
-                        recv_counters_bases[i][cntr]
-                    assert drops == 0, "Detected %d ingress drops" % drops
+            pg_drop_counters = {port_id: sai_thrift_read_pg_drop_counters(
+                self.src_client, port_list['src'][port_id]) for port_id in uniq_srcs}
+            for src_port_id in uniq_srcs:
+                for pg in range(len(pg_drop_counters[src_port_id])):
+                    drops = pg_drop_counters[src_port_id][pg] - pg_drop_counters_bases[src_port_id][pg]
+                    if pg in [3, 4]:
+                        assert drops == 0, "Detected %d lossless drops on PG %d src port %d" % (drops, pg, src_port_id)
+                    elif drops > 0:
```

Comment:
The test pushes memory to the maximum lossless limit, which also passes the ingress drop threshold for lossy traffic due to the memory reserving space for lossless to avoid drops. 
Any random lossy traffic that arrives during this state will thus be dropped under this counter. 

I could set some maximum limit if that's preferrable? 
Due to the flaky and unknown nature of this background traffic, it's also difficult to fix a number. My initial suggestion would be 100, as I recall seeing on the order of 20 drops on an individual port, but with high rarity in this occurring. I typically saw 1-3 drops with about 20% occurrence chance during a test run that would cause the failure.

Reply:
Thanks @rbpittman for your response. I think we can keep the changes as is